src/simulation_engine.cpp

```diff
@@ -1,21 +1,92 @@
 #include <iostream>
 #include <vector>
 #include <memory>
+#include <random>
+#include <algorithm>
+#include <iterator>
 
 #include <simulator/simulation_engine.hpp>
 #include <simulator/simulation_map.hpp>
+#include <simulator/bug.hpp>
 
 namespace simulation {
 
-SimulationEngine::SimulationEngine() {}
+const uint16_t ANTS = 10;
+const uint16_t DOODLEBUGS = 3;
 
-void SimulationEngine::simulate_round() {}
+SimulationEngine::SimulationEngine(): map_(std::make_shared<SimulationMap>()){}
 
-void SimulationEngine::populate_map_with_species() {}
+void SimulationEngine::start_simulation(uint16_t n_rounds) {
+  for(;n_rounds && simulate_round();n_rounds--);
+}
 
-void SimulationEngine::on_bug_born(std::shared_ptr<Bug>) {}
+bool SimulationEngine::simulate_round() {
+  uint16_t species_left = 0;
+  std::cout << "--------" << std::endl;
+  std::cout << "[SimulationEngine] New Round" << std::endl;
+    map_->print_map();
   
-void SimulationEngine::on_bug_died(std::shared_ptr<Bug>) {}
+  // Iterates over cells and makes every bug move/breed/eat/starve accordingly.
+  for(auto cell: map_->get_cells()) {
+    auto bug_ref = cell->get_bug();
+    if(bug_ref && !bug_ref->has_played()) {
```

Comment:
Why do we need to check if it has already played? Can we iterate over the same bug more than once?

Reply:
If we don't, as we're iterating over cells and not over bugs, if a bug moves to a cell that was not yet iterated over, the bug would have two turns in he same round.

Reply:
got it